Первая попытка

In [1]:
from tensorflow import keras
from tensorflow.keras import layers

In [2]:
# Загрузка датасета
(x_train, y_train), (x_test, y_test) = keras.datasets.fashion_mnist.load_data()

In [3]:
# Обработка данных
x_train = x_train.astype("float32") / 255.0
x_test = x_test.astype("float32") / 255.0

In [4]:
# Конвертация меток в формат one-hot encoding
y_train = keras.utils.to_categorical(y_train, num_classes=10)
y_test = keras.utils.to_categorical(y_test, num_classes=10)

In [5]:
# Создание модели
model = keras.Sequential()
model.add(layers.Flatten(input_shape=(28, 28)))
model.add(layers.Dense(128, activation="relu"))
model.add(layers.Dense(64, activation="relu"))
model.add(layers.Dense(10, activation="softmax"))

In [6]:
# Компиляция модели
model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])

In [7]:
# Тренировка модели
model.fit(x_train, y_train, batch_size=128, epochs=10, validation_split=0.2)

Epoch 1/10


375/375 [==============================] - 2s 4ms/step - loss: 0.5727 - accuracy: 0.8004 - val_loss: 0.4428 - val_accuracy: 0.8434
Epoch 2/10
375/375 [==============================] - 1s 3ms/step - loss: 0.4051 - accuracy: 0.8560 - val_loss: 0.4236 - val_accuracy: 0.8462
Epoch 3/10
375/375 [==============================] - 1s 3ms/step - loss: 0.3595 - accuracy: 0.8701 - val_loss: 0.3584 - val_accuracy: 0.8708
Epoch 4/10
375/375 [==============================] - 1s 3ms/step - loss: 0.3306 - accuracy: 0.8799 - val_loss: 0.3475 - val_accuracy: 0.8751
Epoch 5/10
375/375 [==============================] - 1s 3ms/step - loss: 0.3153 - accuracy: 0.8845 - val_loss: 0.3367 - val_accuracy: 0.8796
Epoch 6/10
375/375 [==============================] - 1s 3ms/step - loss: 0.2990 - accuracy: 0.8891 - val_loss: 0.3367 - val_accuracy: 0.8811
Epoch 7/10
375/375 [==============================] - 1s 3ms/step - loss: 0.2887 - accuracy: 0.8941 - val_loss: 0.3355 - val_accuracy: 0.8808
Epoc

In [8]:
# Оцінка точності моделі на тестових даних
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Accuracy: {test_accuracy}")

313/313 [==============================] - 0s 1ms/step - loss: 0.3495 - accuracy: 0.8786
Test Accuracy: 0.878600001335144


Вариант 2 - Добавление регуляризации Dropout

In [9]:
from tensorflow import keras
from tensorflow.keras import layers

# Загрузка датасета
(x_train, y_train), (x_test, y_test) = keras.datasets.fashion_mnist.load_data()

# Обработка данных
x_train = x_train.astype("float32") / 255.0
x_test = x_test.astype("float32") / 255.0

# Конвертация меток в формат one-hot encoding
y_train = keras.utils.to_categorical(y_train, num_classes=10)
y_test = keras.utils.to_categorical(y_test, num_classes=10)

# Создание модели
model = keras.Sequential()
model.add(layers.Flatten(input_shape=(28, 28)))
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dropout(0.5))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dropout(0.5))
model.add(layers.Dense(10, activation='softmax'))

# Компиляция
model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])

# Тренировка
model.fit(x_train, y_train, batch_size=128, epochs=10, validation_split=0.2)

# Оценка
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Accuracy: {test_accuracy}")

Epoch 1/10
375/375 [==============================] - 2s 4ms/step - loss: 0.9871 - accuracy: 0.6456 - val_loss: 0.5260 - val_accuracy: 0.8122
Epoch 2/10
375/375 [==============================] - 1s 3ms/step - loss: 0.6237 - accuracy: 0.7784 - val_loss: 0.4640 - val_accuracy: 0.8340
Epoch 3/10
375/375 [==============================] - 1s 3ms/step - loss: 0.5602 - accuracy: 0.8033 - val_loss: 0.4262 - val_accuracy: 0.8448
Epoch 4/10
375/375 [==============================] - 1s 3ms/step - loss: 0.5187 - accuracy: 0.8205 - val_loss: 0.4028 - val_accuracy: 0.8534
Epoch 5/10
375/375 [==============================] - 1s 3ms/step - loss: 0.4974 - accuracy: 0.8266 - val_loss: 0.3928 - val_accuracy: 0.8568
Epoch 6/10
375/375 [==============================] - 1s 3ms/step - loss: 0.4776 - accuracy: 0.8357 - val_loss: 0.3888 - val_accuracy: 0.8558
Epoch 7/10
375/375 [==============================] - 1s 3ms/step - loss: 0.4662 - accuracy: 0.8383 - val_loss: 0.3914 - val_accuracy: 0.8575
Epoch 

Вариант 3 - добавление дополнительных сверточных слоев

In [10]:
from tensorflow import keras
from tensorflow.keras import layers

# Загрузка датасета
(x_train, y_train), (x_test, y_test) = keras.datasets.fashion_mnist.load_data()

# Обработка данных
x_train = x_train.astype("float32") / 255.0
x_test = x_test.astype("float32") / 255.0

# Конвертация меток в формат one-hot encoding
y_train = keras.utils.to_categorical(y_train, num_classes=10)
y_test = keras.utils.to_categorical(y_test, num_classes=10)

# Создание модели
model = keras.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

# Компиляция
model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])

# Тренировка
model.fit(x_train, y_train, batch_size=128, epochs=10, validation_split=0.2)

# Оценка
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Accuracy: {test_accuracy}")


Epoch 1/10
375/375 [==============================] - 9s 21ms/step - loss: 0.5721 - accuracy: 0.7948 - val_loss: 0.4050 - val_accuracy: 0.8578
Epoch 2/10
375/375 [==============================] - 8s 20ms/step - loss: 0.3639 - accuracy: 0.8700 - val_loss: 0.3373 - val_accuracy: 0.8783
Epoch 3/10
375/375 [==============================] - 8s 20ms/step - loss: 0.3154 - accuracy: 0.8852 - val_loss: 0.3278 - val_accuracy: 0.8824
Epoch 4/10
375/375 [==============================] - 8s 21ms/step - loss: 0.2852 - accuracy: 0.8959 - val_loss: 0.3021 - val_accuracy: 0.8938
Epoch 5/10
375/375 [==============================] - 8s 21ms/step - loss: 0.2618 - accuracy: 0.9051 - val_loss: 0.2783 - val_accuracy: 0.9011
Epoch 6/10
375/375 [==============================] - 8s 21ms/step - loss: 0.2406 - accuracy: 0.9127 - val_loss: 0.2738 - val_accuracy: 0.9027
Epoch 7/10
375/375 [==============================] - 8s 21ms/step - loss: 0.2266 - accuracy: 0.9165 - val_loss: 0.2942 - val_accuracy: 0.893

Вариант 4 - увеличение количества эпох

In [12]:
from tensorflow import keras
from tensorflow.keras import layers

# Загрузка датасета
(x_train, y_train), (x_test, y_test) = keras.datasets.fashion_mnist.load_data()

# Обработка данных
x_train = x_train.astype("float32") / 255.0
x_test = x_test.astype("float32") / 255.0

# Конвертация меток в формат one-hot encoding
y_train = keras.utils.to_categorical(y_train, num_classes=10)
y_test = keras.utils.to_categorical(y_test, num_classes=10)

# Создание модели
model = keras.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

# Компиляция
model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])

# Тренировка
model.fit(x_train, y_train, batch_size=128, epochs=13, validation_split=0.2)

# Оценка
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Accuracy: {test_accuracy}")

Epoch 1/13
375/375 [==============================] - 8s 21ms/step - loss: 0.5801 - accuracy: 0.7900 - val_loss: 0.4077 - val_accuracy: 0.8535
Epoch 2/13
375/375 [==============================] - 8s 21ms/step - loss: 0.3741 - accuracy: 0.8639 - val_loss: 0.3762 - val_accuracy: 0.8625
Epoch 3/13
375/375 [==============================] - 8s 21ms/step - loss: 0.3249 - accuracy: 0.8816 - val_loss: 0.3251 - val_accuracy: 0.8831
Epoch 4/13
375/375 [==============================] - 8s 21ms/step - loss: 0.2938 - accuracy: 0.8942 - val_loss: 0.3051 - val_accuracy: 0.8877
Epoch 5/13
375/375 [==============================] - 8s 21ms/step - loss: 0.2706 - accuracy: 0.9006 - val_loss: 0.3349 - val_accuracy: 0.8816
Epoch 6/13
375/375 [==============================] - 8s 21ms/step - loss: 0.2506 - accuracy: 0.9088 - val_loss: 0.2861 - val_accuracy: 0.8948
Epoch 7/13
375/375 [==============================] - 8s 21ms/step - loss: 0.2316 - accuracy: 0.9141 - val_loss: 0.2782 - val_accuracy: 0.8995

На текущем этапе вариант №4 показал лучшие результаты точости - 91% (Test Accuracy: 0.910099983215332)